dxr/plugins/js/indexers.py

```diff
@@ -0,0 +1,105 @@
+from os.path import basename, dirname, relpath, join, exists
+import json
+import subprocess
+
+from funcy import imap
+
+from dxr.plugins.js.refs import PLUGIN_NAME, QualifiedRef
+from dxr.indexers import (TreeToIndex, FileToIndex,
+                          Extent, Position, iterable_per_line,
+                          with_start_and_end, split_into_lines)
+from dxr.utils import cumulative_sum
+
+
+class ReadAnalysis(object):
+    def __init__(self, tree, lines, contents):
+        self.needles = []
+        self.refs = []
+        # Build map of line number -> byte offset to use for emitting refs.
+        self.offsets = list(cumulative_sum(imap(len, contents.splitlines(True))))
+        for line in lines:
+            row, (start, end) = line['loc']
+            qref = QualifiedRef(tree, (line['sym'], line['name'], line['type']), qualname=line['sym'])
+            if line['kind'] == "use":
+                self.yield_needle(line['type'] + "_ref", row, start, end, line['name'], line['sym'])
+                self.yield_ref(row, start, end, qref)
+            elif line['kind'] == "def":
+                self.yield_needle(line['type'], row, start, end, line['name'], line['sym'])
+                self.yield_ref(row, start, end, qref)
+
+    def yield_ref(self, row, start, end, ref):
+        offset = self.row_to_offset(row)
+        self.refs.append((offset + start, offset + end, ref))
+
+    def row_to_offset(self, line):
+        """Return the byte offset in the file of given line number.
+        """
+        return self.offsets[line - 1]
+
+    def yield_needle(self, filter_name, line, start, end, name, qualname=None):
+        """Add needle for qualified filter_name from line:start
+        to line:end with given name and qualname.
+        """
+        mapping = {'name': name}
+        if qualname:
+            mapping['qualname'] = qualname
+        self.needles.append((PLUGIN_NAME + '_' + filter_name,
+                             mapping,
+                             Extent(Position(row=line, col=start), Position(row=line, col=end))))
```

Comment:
I wonder if JS has any line-continuation stuff that might screw this up in weird cases.